src/detect-engine-threshold.c

```diff
@@ -281,6 +319,99 @@ static inline void RateFilterSetAction(Packet *p, PacketAlert *pa, uint8_t new_a
     }
 }
 
+static int IsThresholdQualify(DetectThresholdEntry* lookup_tsh, const DetectThresholdData *td, uint32_t packet_time)
+{
+    int ret = 0;
+
+    /* Check if we have a timeout enabled, if so,
+    * we still matching (and enabling the new_action) */
+    if (lookup_tsh->tv_timeout != 0) {
+        if ((packet_time - lookup_tsh->tv_timeout) > td->timeout) {
+            /* Ok, we are done, timeout reached */
+            lookup_tsh->tv_timeout = 0;
+        }
+        else {
+            /* Already matching */
+            ret = 1;
+        } /* else - if ((packet_time - lookup_tsh->tv_timeout) > td->timeout) */
+
+    }
+    else {
+        /* Update the matching state with the timeout interval */
+        if ((packet_time - lookup_tsh->tv_sec1) < td->seconds) {
+            lookup_tsh->current_count++;
+            if (lookup_tsh->current_count > td->count) {
+                /* Then we must enable the new action by setting a
+                * timeout */
+                lookup_tsh->tv_timeout = packet_time;
+                ret = 1;
+            }
+        }
+        else {
+            lookup_tsh->tv_sec1 = packet_time;
+            lookup_tsh->current_count = 1;
+        }
+    } /* else - if (lookup_tsh->tv_timeout != 0) */
+
+    return ret;
+}
+
+static void AddEntryToHostStorage(Host *h, DetectThresholdEntry *e, uint32_t packet_time)
+{
+    if (h && e)
+    {
```

Comment:
minor: style: no newline for opening brackets, except for the opening of a function